src/Components/Components/src/Infrastructure/ComponentStatePersistenceManager.cs

```diff
@@ -34,105 +46,94 @@ public ComponentStatePersistenceManager(ILogger<ComponentStatePersistenceManager
 
-        async Task PauseAndPersistState()
-        {
-            State.PersistingState = true;
-            await PauseAsync();
-            State.PersistingState = false;
+            case PersistedStateSerializationMode.WebAssembly:
+                if (_webAssemblyStateIsPersisted)
+                {
+                    throw new InvalidOperationException("State already persisted.");
+                }
+                _webAssemblyStateIsPersisted = true;
+                return PersistStateAsync(store, serializationMode, _webAssemblyCallbacks, _currentWebAssemblyState, dispatcher);
 
-            await store.PersistStateAsync(_currentState);
+            default:
+                throw new InvalidOperationException("Invalid persistence mode");
         }
     }
 
-    internal Task PauseAsync()
+    private Task PersistStateAsync(
+        IPersistentComponentStateStore store,
+        PersistedStateSerializationMode serializationMode,
+        List<Func<Task>> callbacks,
+        Dictionary<string, byte[]> currentState,
+        Dispatcher dispatcher)
     {
-        List<Task>? pendingCallbackTasks = null;
+        return dispatcher.InvokeAsync(PersistStateAsync);
 
-        for (var i = 0; i < _pauseCallbacks.Count; i++)
+        async Task PersistStateAsync()
         {
-            var callback = _pauseCallbacks[i];
-            var result = ExecuteCallback(callback, _logger);
-            if (!result.IsCompletedSuccessfully)
-            {
-                pendingCallbackTasks ??= new();
-                pendingCallbackTasks.Add(result);
-            }
-        }
-
-        if (pendingCallbackTasks != null)
-        {
-            return Task.WhenAll(pendingCallbackTasks);
-        }
-        else
-        {
-            return Task.CompletedTask;
-        }
-
-        static Task ExecuteCallback(Func<Task> callback, ILogger<ComponentStatePersistenceManager> logger)
-        {
-            try
-            {
-                var current = callback();
-                if (current.IsCompletedSuccessfully)
-                {
-                    return current;
-                }
-                else
-                {
-                    return Awaited(current, logger);
-                }
-            }
-            catch (Exception ex)
-            {
-                logger.LogError(new EventId(1000, "PersistenceCallbackError"), ex, "There was an error executing a callback while pausing the application.");
-                return Task.CompletedTask;
-            }
+            State.PersistingState = true;
+            State.SerializationMode = serializationMode;
 
-            static async Task Awaited(Task task, ILogger<ComponentStatePersistenceManager> logger)
+            foreach (var callback in callbacks)
             {
                 try
                 {
-                    await task;
+                    await callback();
```

Comment:
I made the execution of the callbacks sequential here. The reason is that `PersistAsJson` is usually provided in the callback. In order to persist the key, value in the correct dictionary `PersistAsJson` uses the `SerializaionState` property which can be set only here before foreach block. If I'd leave the execution of the callbacks parallel the `SerializaionState` can be set incorrectly.

Reply:
I think running callbacks in parallel should still be possible, as long as we wait for the completion of each callback before mutating `SerializationMode` again.

> If I'd leave the execution of the callbacks parallel the SerializaionState can be set incorrectly.

Do you remember why exactly this was happening?

Reply:
You are right. Changed the code to await the completion of all callbacks.